Fix BN+F.forget

ir5 opened the pull request:
Fix #5555 by simply adding new global config.

TODO:
- [x] Write test
- [x] Fix F.BatchReNormalization as well?
- [x] Write document somewhere

kmaehashi commented:
Jenkins, test this please.

kmaehashi commented on chainer/__init__.py:
```diff
@@ -169,6 +169,7 @@ def is_arrays_compatible(arrays):
     raise TypeError('incorrect dtype name in CHAINER_DTYPE: "{}". '
                     'Only float16/32/64 are allowed.'.format(_chainer_dtype))
 global_config.dtype = numpy.dtype(_chainer_dtype)
+global_config.recompute = False
```
* The configuration key `recompute` sounds too general and a bit confusing to users. How about naming it like `in_recomputation`?
* Could you add the description like this to https://docs.chainer.org/en/latest/reference/configuration.html#configuration-keys?

```rst
This flag is automatically set by :func:`chainer.functions.forget` and not intended to be changed by users.
You can use this flag when implementing your own Link to avoid updating the internal states during recomputation done by `F.forget`.
See the documentation of :func:`chainer.functions.forget` for details.
```

ir5 replied:
Thank you for review. I fixed code.

ir5 commented:
I fixed the problems, so this PR is ready for review.

ir5 commented:
OK, I'll check.

ir5 commented:
I fixed when `self.N` is incremented.
However, it seems that this problem occurs only when `BatchNormalization` takes `finetune=True` argument in `**kwargs`, which we cannot pass to the arguments of `F.forget`. Further, current `F.forget` does not take boolean arguments. So, while I think I have fixed the problem, I didn't write test for it.

toslunar commented:
Could you test it with something like `F.forget(functools.partial(self.link, finetune=True), x)`?

ir5 commented:
Sounds nice. I added test and fixed bug in `BatchRenormalization`.

ir5 commented:
Why Travis CI build failed? It indicates flake8 error at strange point...

toslunar commented:
I restarted Travis after #5764 is merged.

Jenkins, test this please.

toslunar commented:
Jenkins, test this please.

kmaehashi commented:
After discussion, concluded that this issue should be rather considered as bug instead of enhancement.
